### PR TITLE
util/wait: add exponential backoff with context helper

### PR DESCRIPTION
**Pull request by mfojtik**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

In some cases, callers might want to pass `context.Context` to `wait.ExponentialBackoff` to stop on context.Done() (instead of waiting one step).

```dev-release-note
wait.ExponentialBackoffWithContext helper is added for callers that want to cancel retrying when the context passed to this function is done.
```

### CODE REVIEW

**Comment by sttts** [comment]
Lgtm.

/assign @liggitt

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go [hunk 1]
```diff
@@ -280,6 +280,33 @@ func TestExponentialBackoff(t *testing.T) {
...
+	opts := Backoff{Factor: 1.0, Steps: 3}
+
+	// waits up to steps
+	ctx := context.TODO()
+	i := 0
+	err := ExponentialBackoffWithContext(ctx, opts, func(ctx context.Context) (bool, error) {
+		i++
+		return false, nil
+	})
+	if err != ErrWaitTimeout || i != opts.Steps {
```
Comment (yliaog) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go]: opts.Steps is mutated, no?

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go [hunk 2]
```diff
@@ -280,6 +280,33 @@ func TestExponentialBackoff(t *testing.T) {
 	}
 }
 
+func TestExponentialBackoffWithContext(t *testing.T) {
+	opts := Backoff{Factor: 1.0, Steps: 3}
```
Comment (yliaog) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go]: maybe added  case for "Steps: 1", as that is treated specially in the code.

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go [hunk 3]
```diff
@@ -273,6 +278,43 @@ func ExponentialBackoff(backoff Backoff, condition ConditionFunc) error {
...
+		}
+		if backoff.Steps == 1 {
+			break
+		}
+		timer := time.NewTimer(backoff.Step())
+		select {
+		case <-timer.C:
+			continue
+		case <-ctx.Done():
+			return ErrWaitTimeout
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go]: should we call timer.Stop() here?

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go]: a copy of opts is passed in... the original is not modified

**Pull request closed without merging**